src/neon/stage1.rs

```diff
@@ -0,0 +1,531 @@
+    let t3 = vandq_u8(p3, bit_mask);
+    let sum0 = vpaddq_u8(t0, t1);
+    let sum1 = vpaddq_u8(t2, t3);
+    let sum0 = vpaddq_u8(sum0, sum1);
+    let sum0 = vpaddq_u8(sum0, sum0);
+
+    vgetq_lane_u64(vreinterpretq_u64_u8(sum0), 0)
+}
+
+unsafe fn compute_quote_mask(quote_bits: u64) -> u64 {
+    vgetq_lane_u64(
+        vreinterpretq_u64_u8(
+            mem::transmute(vmull_p64(-1, quote_bits as i64))), 0)
+}
+
+struct Utf8CheckingState {
+    has_error: int8x16_t,
+    previous: ProcessedUtfBytes,
+}
+
+impl Default for Utf8CheckingState {
+    #[cfg_attr(not(feature = "no-inline"), inline)]
+    fn default() -> Self {
+        Utf8CheckingState {
+            has_error: vdupq_n_s8(0),
+            previous: ProcessedUtfBytes::default(),
+        }
+    }
+}
+
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn check_ascii_neon(si: &SimdInput) -> bool {
+    let high_bit: uint8x16_t = vdupq_n_u8(0x80);
+    let t0: uint8x16_t = vorrq_u8(si.v0, si.v1);
+    let t1: uint8x16_t = vorrq_u8(si.v2, si.v3);
+    let t3: uint8x16_t = vorrq_u8(t0, t1);
+    let t4: uint8x16_t = vandq_u8(t3, high_bit);
+
+    let v64: uint64x2_t = vreinterpretq_u64_u8(t4);
+    let v32: uint32x2_t = vqmovn_u64(v64);
+    let result: uint64x1_t = vreinterpret_u64_u32(v32);
+
+    vget_lane_u64(result, 0) == 0
+}
+
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn check_utf8(
+    input: &SimdInput,
+    state: &mut Utf8CheckingState,
+) {
+    if check_ascii_neon(input) {
+        // All bytes are ascii. Therefore the byte that was just before must be
+        // ascii too. We only check the byte that was just before simd_input. Nines
+        // are arbitrary values.
+        let verror: int8x16_t =
+            int8x16_t::new(9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 1);
+        state.has_error =
+            vreinterpretq_s8_u8(vorrq_u8(vcgtq_s8(state.previous.carried_continuations, verror),
+                                         vreinterpretq_u8_s8(state.has_error)));
+    } else {
+        // it is not ascii so we have to do heavy work
+        state.previous = check_utf8_bytes(vreinterpretq_s8_u8(input.v0),
+                                          &mut (state.previous), &mut (state.has_error));
+        state.previous = check_utf8_bytes(vreinterpretq_s8_u8(input.v1),
+                                          &mut (state.previous), &mut (state.has_error));
+        state.previous = check_utf8_bytes(vreinterpretq_s8_u8(input.v2),
+                                          &mut (state.previous), &mut (state.has_error));
+        state.previous = check_utf8_bytes(vreinterpretq_s8_u8(input.v3),
+                                          &mut (state.previous), &mut (state.has_error));
+    }
+}
+
+// a straightforward comparison of a mask against input
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn cmp_mask_against_input(input: &SimdInput, m: u8) -> u64 {
+    macro_rules! call {
```

Comment:
What's the story of the `constify_imm8` macro in here, are we missing an intrinsic?